File: python/pycubrid/01_connect.py
"""01_connect.py — Connecting to CUBRID with pycubrid.

Demonstrates:
- Basic connection
- Running a simple query
- Using cursor.description for column metadata
- Connection as context manager
"""

from __future__ import annotations

import pycubrid


def basic_connection() -> None:
    """Connect, query, disconnect."""
    conn = pycubrid.connect(
        host="localhost",
        port=33000,
        database="testdb",
        user="dba",
        password="",
    )

    cursor = conn.cursor()
    cursor.execute("SELECT 1 + 1 AS result")

    row = cursor.fetchone()
    print(f"1 + 1 = {row[0]}")

    cursor.close()
    conn.close()


def connection_info() -> None:
    """Inspect connection metadata."""
    conn = pycubrid.connect(
        host="localhost",
        port=33000,
        database="testdb",
        user="dba",
    )

    cursor = conn.cursor()

    # Server version
    cursor.execute("SELECT version()")
    version = cursor.fetchone()
    print(f"CUBRID version: {version[0]}")

    # Current database
    cursor.execute("SELECT database()")
    db = cursor.fetchone()
    print(f"Database: {db[0]}")

    # Current user
    cursor.execute("SELECT user()")
    user = cursor.fetchone()
    print(f"User: {user[0]}")

    cursor.close()
    conn.close()


def cursor_description() -> None:
    """Use cursor.description to inspect result columns."""
    conn = pycubrid.connect(
        host="localhost",
        port=33000,
        database="testdb",
        user="dba",
    )

    cursor = conn.cursor()
    cursor.execute("SELECT 1 AS id, 'hello' AS name, CAST(3.14 AS DOUBLE) AS value")

    # cursor.description: list of (name, type_code, display_size, internal_size,
    #                               precision, scale, null_ok)
    if cursor.description:
        print("Columns:")
        for col in cursor.description:
            print(f"  {col[0]:10s}  type_code={col[1]}")

    row = cursor.fetchone()
    print(f"Row: {row}")

    cursor.close()
    conn.close()


def multiple_queries() -> None:
    """Run multiple queries on the same connection."""
    conn = pycubrid.connect(
        host="localhost",
        port=33000,
        database="testdb",
        user="dba",
    )

    cursor = conn.cursor()

    queries = [
        "SELECT 1 AS a",
        "SELECT 'hello' AS b",
        "SELECT CURRENT_DATE AS today",
    ]

    for sql in queries:
        cursor.execute(sql)
        row = cursor.fetchone()
        print(f"{sql:40s} → {row[0]}")

    cursor.close()
    conn.close()


if __name__ == "__main__":
    print("=== Basic Connection ===")
    basic_connection()

    print("\n=== Connection Info ===")
    connection_info()

    print("\n=== Cursor Description ===")
    cursor_description()

    print("\n=== Multiple Queries ===")
    multiple_queries()

File: node/cubrid/01_connect.js
// 01_connect.js — Connecting to CUBRID with @cubrid/client.
//
// Demonstrates:
// - Basic connection
// - Running a simple query
// - Connection metadata (version, database, user)
// - Reusing a client for multiple queries

import { createClient } from "@cubrid/client";

const DB_CONFIG = {
  host: "localhost",
  port: 33000,
  database: "testdb",
  user: "dba",
  password: "",
};

async function basicConnection() {
  console.log("=== Basic Connection ===");
  const db = createClient(DB_CONFIG);
  try {
    const rows = await db.query("SELECT 1 + 1 AS result");
    console.log(`1 + 1 = ${rows[0].result}`);
  } finally {
    await db.close();
  }
}

async function connectionInfo() {
  console.log("\n=== Connection Info ===");
  const db = createClient(DB_CONFIG);
  try {
    // Server version
    const [vrow] = await db.query("SELECT version() AS version");
    console.log(`CUBRID version: ${vrow.version}`);

    // Current database
    const [drow] = await db.query("SELECT database() AS db");
    console.log(`Database: ${drow.db}`);

    // Current user
    const [urow] = await db.query("SELECT user() AS u");
    console.log(`User: ${urow.u}`);
  } finally {
    await db.close();
  }
}

async function multipleQueries() {
  console.log("\n=== Multiple Queries ===");
  const db = createClient(DB_CONFIG);
  try {
    const queries = [
      "SELECT 1 AS a",
      "SELECT 'hello' AS b",
      "SELECT CURRENT_DATE AS today",
    ];
    for (const sql of queries) {
      const [row] = await db.query(sql);
      const value = Object.values(row)[0];
      console.log(`${sql.padEnd(35)} → ${value}`);
    }
  } finally {
    await db.close();
  }
}

async function main() {
  await basicConnection();
  await connectionInfo();
  await multipleQueries();
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});
